modules/js-and-css/speculation-rules/class-plsr-url-pattern-prefixer.php

```diff
@@ -86,8 +86,32 @@ public function prefix_path_pattern( string $path_pattern, string $context = 'ho
 	 */
 	public static function get_default_contexts(): array {
 		return array(
-			'home' => trailingslashit( wp_parse_url( home_url( '/' ), PHP_URL_PATH ) ),
-			'site' => trailingslashit( wp_parse_url( site_url( '/' ), PHP_URL_PATH ) ),
+			'home' => self::escape_pattern_string( trailingslashit( wp_parse_url( home_url( '/' ), PHP_URL_PATH ) ) ),
+			'site' => self::escape_pattern_string( trailingslashit( wp_parse_url( site_url( '/' ), PHP_URL_PATH ) ) ),
 		);
 	}
+
+	/**
+	 * Escapes a string for use in a URL pattern component.
+	 * https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param string $str String to be escaped.
+	 * @return string String with backslashes added where required.
+	 */
+	public static function escape_pattern_string( string $str ): string {
+		$replacements = array(
+			'+'  => '\\+',
+			'*'  => '\\*',
+			'?'  => '\\?',
+			':'  => '\\:',
+			'{'  => '\\{',
+			'}'  => '\\}',
+			'('  => '\\(',
+			')'  => '\\)',
+			'\\' => '\\\\',
+		);
+		return strtr( $str, $replacements );
```

Comment:
I wonder whether there's a native PHP function that does this 🤔 

But if not, I think we could simplify this to be a regular expression, since the replacement pattern is always the same (prepend with `\\`):
```suggestion
		// Escape any of these individual characters by prefixing with a backslash.
		return preg_replace( '/[+*?:{}()\\\\]{1}/', '\\\\$0', $str );
```

Reply:
Not that I could find. `addslashes` and `addcslashes` both do it for different sets of special characters, but `strtr` seemed to be the most common approach here.

I can use `preg_replace` (the `{1}` is implied I think but otherwise I'd come up with the same line of code) if you'd prefer.

I'd avoided it because the PHP manual and the results of my Stack Overflow searches both advised not using it where something weaker than regular expressions would suffice, for performance reasons (presumably compiling the pattern has a cost).

I think both are similarly readable but happy to defer to your expertise if `preg_replace` is preferred.

Reply:
The `addcslashes()` function allows you to specify the characters to escape, too. So I think this is better:

```php
return addcslashes( $str, '+*?:{}()\\' );
```

Reply:
You're right! When I first read the manual page I misunderstood the section about how characters that have special C meaning, like \n and \r, and thought it _only_ applied to characters with meaningful C escape sequences. But that's not the case, so `addcslashes` is unambiguously the best solution.